### PR TITLE
COG driver: output exactly square pixels when using -co TILING_SCHEME…

### LINKED ISSUE #5343: TILING_SCHEME=GoogleMapsCompatible generating non-square/unaligned COG's
## Expected behavior and actual behavior.

Expected behaviour:
If a COG is created with TILING_SCHEME=GoogleMapsCompatible, the COG's tiles should be aligned and tiled to the GoogleMapsCompatible Scale Set. (https://gdal.org/drivers/raster/cog.html)

Actual behaviour:
The COG is misaligned, with a non-square pixel resolution and trying to serve directly from this COG requires double the requests to load each tile to match Web Mercator.

Originally from: https://github.com/developmentseed/titiler/issues/431

## Steps to reproduce the problem.
http://static2.skysight.io/himawari.tiff is a sample file (GEOS projected satellite imagery)

```
gdal_translate himawari.tiff himawari-googlemaps.tiff -of COG -co TILING_SCHEME=GoogleMapsCompatible -co COMPRESS=JPEG -co NUM_THREADS=ALL_CPUS -co ZOOM_LEVEL_STRATEGY=UPPER
```

Observe the non-square pixel size:
Pixel Size = (1222.992452562820063,-1222.992452562820290)

```
gdalinfo himawari-googlemaps.tiff 
Driver: GTiff/GeoTIFF
Files: himawari-googlemaps.tiff
Size is 32768, 25088
Coordinate System is:
PROJCRS["WGS 84 / Pseudo-Mercator",
    BASEGEOGCRS["WGS 84",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["Popular Visualisation Pseudo-Mercator",
        METHOD["Popular Visualisation Pseudo Mercator",
            ID["EPSG",1024]],
        PARAMETER["Latitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["False easting",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["easting (X)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["northing (Y)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["Web mapping and visualisation."],
        AREA["World between 85.06°S and 85.06°N."],
        BBOX[-85.06,-180,85.06,180]],
    ID["EPSG",3857]]
Data axis to CRS axis mapping: 1,2
Origin = (-20037508.342789243906736,15341217.324948014691472)
Pixel Size = (1222.992452562820063,-1222.992452562820290)
Metadata:
  AREA_OR_POINT=Area
  TIFFTAG_DATETIME=2022:02:05 00:00:20
Image Structure Metadata:
  COMPRESSION=YCbCr JPEG
  INTERLEAVE=PIXEL
  JPEGTABLESMODE=1
  JPEG_QUALITY=75
  LAYOUT=COG
  SOURCE_COLOR_SPACE=YCbCr
Tiling Scheme:
  NAME=GoogleMapsCompatible
  ZOOM_LEVEL=7
Corner Coordinates:
Upper Left  (-20037508.343,15341217.325) (180d 0' 0.00"W, 79d41'13.86"N)
Lower Left  (-20037508.343,-15341217.325) (180d 0' 0.00"W, 79d41'13.86"S)
Upper Right (20037508.343,15341217.325) (180d 0' 0.00"E, 79d41'13.86"N)
Lower Right (20037508.343,-15341217.325) (180d 0' 0.00"E, 79d41'13.86"S)
Center      (   0.0000000,  -0.0000000) (  0d 0' 0.01"E,  0d 0' 0.00"S)
Band 1 Block=256x256 Type=Byte, ColorInterp=Red
  NoData Value=0
  Overviews: 16384x12544, 8192x6272, 4096x3136, 2048x1568, 1024x784, 512x392, 256x196
Band 2 Block=256x256 Type=Byte, ColorInterp=Green
  NoData Value=0
  Overviews: 16384x12544, 8192x6272, 4096x3136, 2048x1568, 1024x784, 512x392, 256x196
Band 3 Block=256x256 Type=Byte, ColorInterp=Blue
  NoData Value=0
  Overviews: 16384x12544, 8192x6272, 4096x3136, 2048x1568, 1024x784, 512x392, 256x196
```

## Operating system
Ubuntu 20.04
## GDAL version and provenance
GDAL 3.3.2, released 2021/09/01

### PR DESCRIPTION
**Pull request by rouault**
… (fixes #5343)

and emit warning messages when using RES or EXTENT options in that mode.

### CODE REVIEW

**Comment by rouault** [comment]
I'm not really sure about the "fixing" part of the commit message. See https://github.com/OSGeo/gdal/issues/5343#issuecomment-1044544565

File: alg/gdalwarpoperation.cpp [hunk 1]
```diff
@@ -2970,7 +2985,7 @@ CPLErr GDALWarpOperation::ComputeSourceWindow(
 #if DEBUG_VERBOSE
     CPLDebug(
         "WARP",
-        "dst=(%d,%d,%d,%d) raw src=(minx=%.8g,miny=%.8g,maxx=%.8g,maxy=%.8g)",
+        "dst=(%d,%d,%d,%d) raw src=(minx=%.18g,miny=%.18g,maxx=%.18g,maxy=%.18g)",
```
Comment (vincentsarago) [inline alg/gdalwarpoperation.cpp:2988]: 😍

Comment (rouault) [reply alg/gdalwarpoperation.cpp:2988]: don't get too excited by that: that's just a debug trace not enabled in normal builds :-)

Comment (vincentsarago) [reply alg/gdalwarpoperation.cpp:2988]: 😂

**Comment by rouault** [comment]
@vincentsarago Do you believe this is worth backporting to GDAL 3.4 branch ? or would you have a workaround in titiler itself for older GDAL versions anyway ?

**Comment by vincentsarago** [comment]
well to me everything is worth back porting to GDAL 3.4 (mostly because rasterio wheels are built with gdal 3.4), but I don't want to create much trouble for you.

**Pull request merged**